### PR TITLE
Make errors nicer when TDNR fails

### LINKED ISSUE #4632: Confusing TDNR related error message
I was working on the Share SDK library and some JSON Decoders and ran into this error message:

https://github.com/unisonweb/unison/assets/2371/f9dba78b-99bb-48a1-a5c2-d0e9f7d94eb0

Its confusing on a series of fronts:

1. `text` worked before, and now doesn't. The real issue is that `or` isn't qualified, but this is reported via `text`
2. There's so much output on the screen that my eyes just glaze over. The message is about as effective as a message that just said "error" (meaning not effective).
3. The project guts are showing. We have a few cases where this is still the case and I think a filter on output messages could help a lot on this area. I know we plan to a different storage mechanism for projects, but in the meantime its really hurting the usability of these error messages

### PR DESCRIPTION
**Pull request by runarorama**
Fixes #4632

This makes the error message for TDNR failures simpler and clearer.

* Removes references to the namespace, since we're usually in the project root and UCM tells you which namespace you're in anyway.
* Infers the likely cause a little better and doesn't suggest e.g. a totally missing symbol when there are TDNR suggestions available.
* Added "TDNR masking the true error" to the list of common causes.
* Removed defunct instructions for adding a library.
* Added clear instructions for resolving the error
* Suffixified the TDNR suggestions.
* Removed bullets from TDNR suggestion list.
* Improved wrapping.
* Sorted suggestions by the length of the FQN, as a proxy for relevance.

Example of new behaviour:

image

### CODE REVIEW

File: parser-typechecker/src/Unison/PrintError.hs [hunk 1]
```diff
@@ -173,7 +173,7 @@ renderTypeError ::
   String ->
   Path.Absolute ->
   Pretty ColorText
-renderTypeError e env src curPath = case e of
+renderTypeError e env src _curPath = case e of
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrintError.hs]: Are you able to remove this arg entirely please?

File: parser-typechecker/src/Unison/PrintError.hs [hunk 2]
```diff
@@ -629,51 +629,68 @@ renderTypeError e env src curPath = case e of
         _ -> Pr.wrap $ "It should be of type " <> Pr.group (style Type1 (renderType' env expectedType) <> ".")
   UnknownTerm {..} ->
     let (correct, wrongTypes, wrongNames) =
-          foldr sep id suggestions ([], [], [])
-        sep (C.Suggestion name typ _ match) r =
+          foldr sep id (sortOn (Text.length . C.suggestionName) suggestions) ([], [], [])
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrintError.hs]: Is this meant to sort by the length of the FQN?
If you're aiming to select the most relevant names you might want to count the number of `.` separated segments instead.

File: parser-typechecker/src/Unison/PrintError.hs [hunk 3]
```diff
@@ -629,51 +629,68 @@ renderTypeError e env src curPath = case e of
...
+        sep s@(C.Suggestion _ _ _ match) r =
           case match of
-            C.Exact -> (_1 %~ ((name, typ) :)) . r
-            C.WrongType -> (_2 %~ ((name, typ) :)) . r
-            C.WrongName -> (_3 %~ ((name, typ) :)) . r
-        libPath = Path.absoluteToPath' curPath Path.:> "lib"
+            C.Exact -> (_1 %~ (s :)) . r
+            C.WrongType -> (_2 %~ (s :)) . r
+            C.WrongName -> (_3 %~ (s :)) . r
+        -- libPath = Path.absoluteToPath' curPath Path.:> "lib"
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/PrintError.hs]:
```suggestion
```

File: unison-src/transcripts-manual/rewrites.output.md [hunk 4]
```diff
@@ -306,21 +306,17 @@ Instead, it should be an unbound free variable, which doesn't typecheck:
 
   Loading changes detected in scratch.u.
 
-  I couldn't find any definitions matching the name bar21 inside the namespace .
+  I couldn't figure out what bar21 refers to here:
   
      19 |   bar21
   
+  Its type should conform to _ .
```
Comment (ChrisPenner) [inline unison-src/transcripts-manual/rewrites.output.md]: What's up with the type `_` here? Doesn't seem right 🤔

File: unison-src/transcripts/destructuring-binds.output.md [hunk 5]
```diff
@@ -86,21 +86,17 @@ ex4 =
 
   Loading changes detected in scratch.u.
 
-  I couldn't find any definitions matching the name a inside the namespace .
+  I couldn't figure out what a refers to here:
   
       2 |   (a,b) = (a Nat.+ b, 19)
   
+  Its type should conform to Nat .
```
Comment (ChrisPenner) [inline unison-src/transcripts/destructuring-binds.output.md]: Maybe either remove the `.` or drop the space :)

**Review by ChrisPenner — COMMENTED** [review]
Looks good! I'm curious what's up with the transcripts outputting `_` as the expected type though 🤔

File: parser-typechecker/src/Unison/PrintError.hs [hunk 6]
```diff
@@ -629,51 +629,68 @@ renderTypeError e env src curPath = case e of
...
+                      Type.Var' (TypeVar.Existential {}) -> Pr.wrap "its type could be anything." <> "\n"
+                      _ ->
+                        Pr.wrap
+                          "no term with that name would pass typechecking. I think its type should be:"
+                          <> "\n\n"
+                          <> Pr.indentN 4 (style Type1 (renderType' env expectedType))
+                          <> "\n\n"
+                          <> Pr.wrap "If that's not what you expected, you may have a type error somewhere else in your code."
+                    <> " "
+                    <> Pr.wrap (Pr.bold "Help me out by using a more specific name here or adding a type annotation.")
```
Comment (aryairani) [inline parser-typechecker/src/Unison/PrintError.hs]:
```suggestion
                    <> Pr.wrap ("Help me out by" <> Pr.bold "using a more specific name here" <> "or" <> Pr.bold "adding a type annotation.")
```
Maybe less bold, to draw attention more directly?

File: unison-src/transcripts/destructuring-binds.output.md (hunk 5 above)
Comment (runarorama) [reply unison-src/transcripts/destructuring-binds.output.md]: This is `wrap` adding spaces again

File: unison-src/transcripts-manual/rewrites.output.md (hunk 4 above)
Comment (runarorama) [reply unison-src/transcripts-manual/rewrites.output.md]: Hmm, used to be "there are no constraints on its type". I must have botched the pattern match on the type here.

**Comment by aryairani** [comment]
Currently a compile error in `parser-typechecker/tests/Unison/Test/UnisonSources.hs`

**Comment by aryairani** [comment]
> Looks good! I'm curious what's up with the transcripts outputting `_` as the expected type though 🤔

@runarorama I have this question too — It looks like it used to say 
> There are no constraints on its type.

but now it says 
> Its type should conform to _ 

(and missing the `.`)

Otherwise, I'm eager to get it merged before transcripts start failing again 😅

**Comment by runarorama** [comment]
`Its type should conform to _` is a regression in trunk unrelated to this change.

I removed the `.` after the type because `Foo.Bar -> Baz.Qux` reads better than `Foo.Bar -> Baz.Qux.`

**Comment by aryairani** [comment]
> `Its type should conform to _` is a regression in trunk unrelated to this change.

Hm, but I don't see that output in the transcripts on trunk?

https://github.com/unisonweb/unison/blob/trunk/unison-src/transcripts/resolution-failures.output.md?plain=1#L127
```
There are no constraints on its type.
```

vs

https://github.com/unisonweb/unison/blob/runarorama/improve-tdnr-message/unison-src/transcripts/resolution-failures.output.md?plain=1#L121
```
Its type should conform to:
  
      _
```

I think it might be due to losing [this logic](https://github.com/unisonweb/unison/pull/4641/files#diff-ea0df981901f1840e463f80adb91c04462f7f4e0b7978c159578e1467be88f60L658-L663), which is present [here](https://github.com/unisonweb/unison/pull/4641/files#diff-ea0df981901f1840e463f80adb91c04462f7f4e0b7978c159578e1467be88f60R679-R680), but missing [here](https://github.com/unisonweb/unison/pull/4641/files#diff-ea0df981901f1840e463f80adb91c04462f7f4e0b7978c159578e1467be88f60R651) and [here](https://github.com/unisonweb/unison/pull/4641/files#diff-ea0df981901f1840e463f80adb91c04462f7f4e0b7978c159578e1467be88f60R651).

> I removed the `.` after the type because `Foo.Bar -> Baz.Qux` reads better than `Foo.Bar -> Baz.Qux.`

Ok 👍

**Comment by runarorama** [comment]
OK, I believe I have addressed all of the issues and fixed the places where it was missing the existentials check. Output looks good in all transcripts. 🚢

**Review by aryairani — APPROVED** [review]
Great!

**Pull request merged**